diff --git a/core/uwsgi.c b/core/uwsgi.c
index 0c4dfa5..da9e308 100644
--- a/core/uwsgi.c
+++ b/core/uwsgi.c
@@ -2231,7 +2231,7 @@ void uwsgi_setup(int argc, char *argv[], char *envp[]) {
 		uwsgi_opt_flock(NULL, uwsgi.flock_wait2, NULL);
 
 	// setup master logging
-	if (uwsgi.log_master && !uwsgi.daemonize2 && uwsgi.logfile)
+	if (uwsgi.log_master && !uwsgi.daemonize2 && (uwsgi.logfile || !uwsgi.logto2))
 		uwsgi_setup_log_master();
 
 	// setup offload engines
