diff --git a/core/uwsgi.c b/core/uwsgi.c
index 4ff6e31..76631ab 100644
--- a/core/uwsgi.c
+++ b/core/uwsgi.c
@@ -2198,7 +2198,7 @@ void uwsgi_setup(int argc, char *argv[], char *envp[]) {
 		uwsgi_opt_flock(NULL, uwsgi.flock_wait2, NULL);
 
 	// setup master logging
-	if (uwsgi.log_master && !uwsgi.daemonize2)
+	if (uwsgi.log_master && !uwsgi.daemonize2 && !uwsgi.logto2)
 		uwsgi_setup_log_master();
 
 	// setup offload engines
@@ -2481,6 +2481,7 @@ int uwsgi_start(void *v_argv) {
 	if (uwsgi.logto2) {
 		if (!uwsgi.is_a_reload || uwsgi.log_reopen) {
 			logto(uwsgi.logto2);
+			uwsgi_setup_log_master();
 		}
 	}
 
