diff --git a/core/ssl.c b/core/ssl.c
index 8fa519f..a52f318 100644
--- a/core/ssl.c
+++ b/core/ssl.c
@@ -243,6 +243,7 @@ SSL_CTX *uwsgi_ssl_new_server_context(char *name, char *crt, char *key, char *ci
                 DH *dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
                 BIO_free(bio);
                 if (dh) {
+                        SSL_CTX_set_options(ctx, SSL_OP_SINGLE_DH_USE);
                         SSL_CTX_set_tmp_dh(ctx, dh);
                         DH_free(dh);
                 }
@@ -252,6 +253,7 @@ SSL_CTX *uwsgi_ssl_new_server_context(char *name, char *crt, char *key, char *ci
 #ifdef NID_X9_62_prime256v1
         EC_KEY *ecdh = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
         if (ecdh) {
+                SSL_CTX_set_options(ctx, SSL_OP_SINGLE_ECDH_USE);
                 SSL_CTX_set_tmp_ecdh(ctx, ecdh);
                 EC_KEY_free(ecdh);
         }
